R/abs_api.R

```diff
@@ -201,17 +216,77 @@ abs_api_url <- function(path, query = NULL) {
   out
 }
 
+
+#' Construct a datakey
+#'
+#' @param datakey A named list matching variables to codes
+#' @param datastructure A datastructure retrieved with `abs_datastructure()`
+#'
+#' @return A datakey string
+#' @noRd
+#' @keywords internal
+#'
+#' @examples
+#' z <- abs_datastructure("ERP_COB")
+#' abs_api_match_key(list(SEX=1:3, REGION="AUS"), z)
+#' abs_api_match_key(list(SEX=1:10, REGION="AUS"), z)
+#' try(abs_api_match_key(list(SX=1:3, REGION="AUS"), z))
+#' try(abs_api_match_key(list(UNIT_MEASURE="AUD", REGION="AUS"), z))
+abs_api_match_key <- function(datakey, datastructure) {
+  stopifnot("`datakey` must be a named list" = is.list(datakey))
+  stopifnot("`datakey` must be a named list" = length(names(datakey))==length(datakey))
+
+  bad_vars <- setdiff(names(datakey), datastructure$var)
+  if (length(bad_vars) > 0) {
+    stop(call. = FALSE,
+         "Variable(s) `", paste(bad_vars, collapse = "`, `"),
+         "` not found. Please review the datastructure with `abs_datastructure()`"
+    )
+  }
+
+  key <- purrr::imap_dfr(datakey, function(vals, var) {
+    ds <- datastructure[datastructure$var %in% var, ]
+
+    bad_vals <- setdiff(vals, ds$code)
+    if (length(bad_vals) > 0) {
+      if (length(bad_vals)>5) bad_vals <- c(bad_vals[1:5], "...")
+      warning(call. = FALSE,
+              "Variable `", var, "` does not contain codes: ",
+              paste(bad_vals, collapse=", ")
+      )
+    }
+
+    if (any(is.na(ds$position))) {
+      stop(call. = FALSE,
+           "Cannot filter on `", var,
+           "`. Please review the datastructure with `abs_datastructure()`"
+      )
+    }
+
+    ds[ds$code %in% vals, ]
+  })
+
+
+  pos <- unique(datastructure$position)
+  pos <- sort(pos[!is.na(pos)])
```

Comment:
`sort` removes NAs by default so `pos <- sort(pos)` is sufficient (and avoids two passes).

Reply:
Thanks!